drivers/ad5770r/ad5770r.c

```diff
@@ -0,0 +1,658 @@
+ */
+int32_t ad5770r_set_channel_en(struct ad5770r_dev *dev,
+			       const struct ad5770r_channel_switches *channel_enable)
+{
+	int32_t ret;
+
+	ret = ad5770r_spi_reg_write(dev,
+				    AD5770R_CH_ENABLE,
+				    AD5770R_CH_ENABLE_SET(channel_enable->en0, AD5770R_CH0) |
+				    AD5770R_CH_ENABLE_SET(channel_enable->en1, AD5770R_CH1) |
+				    AD5770R_CH_ENABLE_SET(channel_enable->en2, AD5770R_CH2) |
+				    AD5770R_CH_ENABLE_SET(channel_enable->en3, AD5770R_CH3) |
+				    AD5770R_CH_ENABLE_SET(channel_enable->en4, AD5770R_CH4) |
+				    AD5770R_CH_ENABLE_SET(channel_enable->en5, AD5770R_CH5));
+
+	if (ret)
+		return ret;
+
+	dev->channel_enable = *channel_enable;
+
+	return ret;
+}
+
+/**
+ * Get status value.
+ * @param dev - The device structure.
+ * @param mon_setup - The structure that contains the monitor setup
+ *		values.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_get_status(struct ad5770r_dev *dev,
+			   uint8_t *status)
+{
+	return ad5770r_spi_reg_read(dev,
+				    AD5770R_STATUS,
+				    status);
+};
+
+/**
+ * Set the channel monitor configuration.
+ * @param dev - The device structure.
+ * @param mon_setup - The structure that contains the monitor setup
+ *		values.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_monitor_setup(struct ad5770r_dev *dev,
+				  const struct ad5770r_monitor_setup *mon_setup)
+{
+	int32_t ret;
+
+	ret = ad5770r_spi_reg_write(dev,
+				    AD5770R_MONITOR_SETUP,
+				    AD5770R_MONITOR_SETUP_MON_CH(mon_setup->monitor_channel) |
+				    AD5770R_MONITOR_SETUP_IB_EXT_EN(mon_setup->ib_ext_en) |
+				    AD5770R_MONITOR_SETUP_MUX_BUFFER(mon_setup->mux_buffer) |
+				    AD5770R_MONITOR_SETUP_MON_FUNCTION(mon_setup->monitor_function));
+
+	if (ret)
+		return ret;
+
+	dev->mon_setup = *mon_setup;
+
+	return ret;
+};
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ *		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_init(struct ad5770r_dev **device,
+		     const struct ad5770r_init_param *init_param)
+{
+	struct ad5770r_dev	*dev;
+	uint8_t product_id_l, product_id_h;
+	int32_t ret;
+	enum ad5770r_channels i;
+
+	dev = (struct ad5770r_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+
+	/* Query device presence */
+	ad5770r_spi_reg_read(dev, AD5770R_PRODUCT_ID_L, &product_id_l);
+	ad5770r_spi_reg_read(dev, AD5770R_PRODUCT_ID_H, &product_id_h);
+
+	if (product_id_l != 0x04 || product_id_h != 0x40) {
+		printf("failed to read id (0x%X : 0x%X)\n", product_id_l,
+		       product_id_h);
+		return FAILURE;
+	}
+
+	/* Device settings */
+	ret = ad5770r_set_device_spi(dev,
+				     &init_param->dev_spi_settings);
```

Comment:
there is no case to handle  `init_param` being null;
there are 2 ways to do it:
1) if `init_param` is NULL, return -EINVAL
2) if `init_param` is NULL pass a default static `init_param` with default values for this device

your preference which one to choose